JS.md

```diff
@@ -0,0 +1,237 @@
+# JavaScript API
+
+In the [MVP](MVP.md), the only way to access WebAssembly on the Web is through
+an explicit JS API which is defined below. (In the future, WebAssembly may also
+be loaded and run directly from an HTML `<script type='module'>` tag—and
+any other Web API that loads ES6 modules via URL—as part of 
+[ES6 Module integration](Modules.md#integration-with-es6-modules).)
+
+## The `Wasm` object
+
+The `Wasm` object is the initial value of the `Wasm` property of the global
+object. Like the `Math` and `JSON` objects, the `Wasm` object is a plain
+JS object (not a constructor or function) that acts like a namespace and has the
+following properties:
+
+### Constructor Properties of the `Wasm` object
+
+* `Wasm.Module` : see [`Wasm.Module` objects](#wasmmodule-objects)
+* `Wasm.Instance` : see [`Wasm.Instance` objects](#wasminstance-objects)
+* `Wasm.CompileError` : a [NativeError](http://tc39.github.io/ecma262/#sec-nativeerror-object-structure)
+   which indicates an error during WebAssembly decoding or validation
+* `Wasm.RuntimeError` : a [NativeError](http://tc39.github.io/ecma262/#sec-nativeerror-object-structure)
+   which indicates an error while running WebAssembly code
+
+### Function Properties of the `Wasm` object
+
+#### `Wasm.compile`
+
+The `compile` function has the signature:
+```
+Promise<Wasm.Module> compile(TypedArray bytes)
+```
+If the given `bytes` argument is not a
+[TypedArray object](http://tc39.github.io/ecma262/#sec-typedarray-objects),
+the returned `Promise` is [rejected](http://tc39.github.io/ecma262/#sec-rejectpromise)
+with a `TypeError`.
+
+Otherwise, this function starts an asychronous task to compile a `Wasm.Module`
+as described in the [`Wasm.Module` constructor](#wasmmodule-constructor).
+On success, the `Promise` is [fulfilled](http://tc39.github.io/ecma262/#sec-fulfillpromise)
+with the resulting `Wasm.Module` instance. On failure, the `Promise` is 
+[rejected](http://tc39.github.io/ecma262/#sec-rejectpromise) with the 
+`Wasm.CompileError`.
+
+The asynchronous compilation is logically performed on a copy of the state of
+the given TypedArray captured during the call to `compile`; subsequent mutations
+of the TypedArray after `compile` return do not affect ongoing compilations.
+
+In the [future](FutureFeatures.md#streaming-compilation), this function can be
+extended to accept a [stream](https://streams.spec.whatwg.org), thereby enabling
+asynchronous, background, streaming compilation.
+
+## `Wasm.Module` Objects
```

Comment:
Ok, @jonco3 helped me understand these concepts a bit more.  It seems like `Wasm.Instance` directly corresponds to a [Module Record](http://tc39.github.io/ecma262/#sec-abstract-module-records) which is a separate-but-related thing from the Module Namespace object.  Since the Module Namespace object's structure is super-fixed (it's only exports), if Module Records were to be one day reflected as first-class JS objects (they are already represented this way internally in SM which is useful for self-hosting and unit testing), then they'd be a distinct JS object from the Namespace object.

So that suggests keeping exports and instance separate as it is now and changing `exports`s value from a POJO to a Module Namespace Exotic Object.  Talking to @jonco3, this would be pretty easy to implement in SM, we think.  In the spec, Module Namespace Objects have a [[Module]] slot, so we'd need to define a Module Record for each `Wasm.Instance` which seems doable...

That reopens the question is whether to keep `instantiate` a method or morph it into the `Wasm.Instantiate` constructor.